Add solution design document to encoded token feature (#1740)
cyberark/conjur: design/proposals/encoded_token_response.md

# Encoded Token Response - Solution Design

## Table of Contents
  * [Table of Contents](#table-of-contents)
  * [Glossary](#glossary)
  * [Useful links](#useful-links)
  * [Background](#background)
  * [Issue description](#issue-description)
  * [Solution - Accept HTTP Header](#solution---accept-http-header)
    + [Design](#design)
    + [Backwards compatibility](#backwards-compatibility)
  * [Solution - URL Patten](#solution---url-patten)
    + [Design](#design-1)
    + [Backwards compatibility](#backwards-compatibility-1)
  * [Preferred Solution](#preferred-solution)
  * [Security](#security)
  * [Test Plan](#test-plan)
  * [Documentation](#documentation)
  * [Open questions](#open-questions)

## Glossary

| **Term** | **Description** |
|----------|-----------------|

> | encoded         | base64 encoded               |

Reviewer: - Flattened JWS JSON Serialization
- Encoded JWS Compact Serialization

## Useful links
1. [HTTP Content Negotiation Best Practices](https://restfulapi.net/content-negotiation/)
2. MDN web docs:
    1. [HTTP Accept Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept)
    2. [HTTP Accept-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding)
    3. [HTTP Content-Type Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)
    4. [HTTP Content-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding)
3. [Rails Routing Guides](https://guides.rubyonrails.org/routing.html#route-globbing-and-wildcard-segments)
4. [Correct Usage of JSON](https://stackoverflow.com/questions/477816/what-is-the-correct-json-content-type)

## Background

> When a user connects to Conjur (via UI or CLI etc.), Conjur uses `authn` authenticator to perform the authentication.

Reviewer: ```suggestion
As a Conjur User or Host
I want to send authentication request to `authn` authenticator
So that I recieve an encoded access token 
```

Reviewer: We try to have a background structed like such:
“As a [persona], I [want to], [so that].”
https://www.atlassian.com/agile/project-management/user-stories

Author: I receive a JWT access token - not encoded one. 
Changed

The current situation is:

![alt text](encoded_token_chart.jpeg "Login process")

> The response is a JSON like:

Reviewer: ```suggestion
Current authentication response is a Flattened JWS JSON Serialization, e.g:
```

Author: Changed

```json
{

>     "protected": "eyJhbGciOiJjb25qdXIub3JnL3Nsb3NpbG8vdjIiLCJraWQiOiJjNjEyNTgwNzQ1YjkxMTBjNGQ5YjBkZWFkYTZkMGU0NjlmNDNkMjE5ZDIwZjdiMGM1YmNlMDVmYjE0MGYzNGM4In0=",
>     "payload": "eyJzdWIiOiJhZG1pbiIsImlhdCI6MTU5NjEyNzg1N30=",
>     "signature": "CZgBa6SwfumtC90QoIIh_vfeXgO_I1SfFYrl4RMiCrW_NtMPOoBIbV6WKze1_P6EjnGAsOHoZmEJ3g_7GpTAAuFU6arj2Ku4JDUup--jPTvfAcPxZViQBISTTmwLUXlDqPsIFbG9lfLV-cCM0v_6BbU0dgWba-w-k41PGHQupioSe_kr5b6izo9prSYlCpn_s9GZz41AYDPxiEPZHk-Xe2fjJCRt5vUI_QJKGA8pScMUozFIxXeGeH3nM28W54rqdRQoJxvgW7YUH_5x0nNurLgtJQSZHI38x9uTassLiRluj5AtkB6DCPCvRQiNdq2yyAQOZuVe7Cj8baHYDDrIpJNUjW2VeJ22f-vcGrsoB6X3H7Y2hfzBxw1NQjEaDaz5"

Reviewer: ```suggestion
    "protected": "ey...",
    "payload": "eyJ...",
    "signature": "CZ..."
```

Author: Changed

}
``` 

> In order to work with the server the user **must** encode the response:

Reviewer: ```suggestion
In order to use this token for restricted endpoint it **must** be encode and added to `Authorization: Token token=` header:
```

Author: Changed

```sh
echo "Get Conjur access token using an Azure access token"
# Get an authn-azure Conjur access token for host azure-apps/test-app
authn_azure_response=$(curl -k -X POST \
  -H "Content-Type: application/x-www-form-urlencoded" \
  --data "jwt=$azure_access_token" \
  https://"$CONJUR_SERVER_DNS":8443/authn-azure/test/cucumber/host%2Fazure-apps%2Ftest-app/authenticate)
authn_azure_access_token=$(echo -n "$authn_azure_response" | base64 | tr -d '\r\n')

echo "Retrieve a secret using the Conjur access token"
# Retrieve a Conjur secret using the authn-azure Conjur access token
secret=$(curl -k -H "Authorization: Token token=\"$authn_azure_access_token\"" \
  https://"$CONJUR_SERVER_DNS":8443/secrets/cucumber/variable/secrets/test-variable)

echo "Retrieved secret ${secret} from Conjur!!!"
```

> Taken from: https://github.com/cyberark/conjur/blob/application_identity_validation/ci/authn-azure/run-authn-azure.sh#L16

Reviewer: ```suggestion
[Code reference]( https://github.com/cyberark/conjur/blob/application_identity_validation/ci/authn-azure/run-authn-azure.sh#L16)
```

Author: Chnaged

> The required situation is to support both `application/json` response (as above) and encoded one.

Reviewer: ```suggestion
The requested situation is to support both response of current **Flattened** JWS JSON Serialization with content type `application/json` and encoded JWS **Compact** Serialization with content type `text\plain`
```

Author: Changed

The user will have the option to receive an encoded response which skips the encoding process afterwards. 

## Issue description
1. HTTP server best practices: Using `application\json` as the content type to encoded output (not a JSON) is bad practice.
    So encoded output shouldn't use `application/json` unless it will be in JSON format. 
    The `../autheticate` response is in JSON format so it makes sense to use `application\json` as the response content type.
        As well as `../logic`,  which returns the api key as `text\plain`.

## Solution - Accept HTTP Header
According to [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept):
 > The `Accept` request HTTP header advertises which content types the client is able to understand. Using content negotiation, the server then selects one of the proposals, uses it and informs the client of its choice with the `Content-Type` response header.

To conclude, we can choose the response's `Content-Type` by the request's `Accept` header.

### Design
When `../authenticate` request the response will be encoded **only** if the `Accept` header is `text/plain`. 

> Otherwise, returns decoded json access token as `application/json`.

Reviewer: ```suggestion
Otherwise, returns json access token as `application/json`.
```

Author: Changed

```ruby
def authenticate
    # get the authn access token
    content_type = (request.headers["Accept"] == "text\plain") ? :text : :json
    # encode Base64 if needed
    render content_type => auth_token

> ```

Reviewer: Can be removed from design doc - since it might change

Author: PSEUDO code

### Backwards compatibility
Default behaviour: Only if the `Accept` is set to `text\plain` the response will be encoded.

## Solution - URL Patten
According to [Rails Routing Guides](https://guides.rubyonrails.org/routing.html#route-globbing-and-wildcard-segments)  we can use wildcards to suggests the requested `Content_type`.

By requesting `../authenticate.encoded` the output will be encoded access token.    

### Design
When `../authenticate` request the response will be encoded **only** if the url suffix (wildcard) is `.encoded`.

> Otherwise, returns decoded json access token as `application/json`

Reviewer: ```suggestion
Otherwise, returns json access token as `application/json`
```

Author: Changed

  * [Documentation](#documentation)
  * [Open questions](#open-questions)

## Glossary

| **Term** | **Description**  |
|----------|------------------|
| encoded  |base64 encoded    |
| JWS      |json web signature|
| JWT      |json web token    |

## Useful links
1. [HTTP Content Negotiation Best Practices](https://restfulapi.net/content-negotiation/)
2. MDN web docs:
    1. [HTTP Accept Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept)
    2. [HTTP Accept-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding)
    3. [HTTP Content-Type Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)
    4. [HTTP Content-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding)
3. [Rails Routing Guides](https://guides.rubyonrails.org/routing.html#route-globbing-and-wildcard-segments)
4. [Correct Usage of JSON](https://stackoverflow.com/questions/477816/what-is-the-correct-json-content-type)

## Background

As a Conjur User or Host, I want to send authentication request to `authn` authenticator (for example).
So that I receive a JWT access token.

> 

Reviewer: Comment on image from meeting:
1. Remove login step
2. Change authorize to authenticate
3. http(s) for both DAP and Conjur communication

Author: Added

echo "Retrieve a secret using the Conjur access token"
# Retrieve a Conjur secret using the authn-azure Conjur access token
secret=$(curl -k -H "Authorization: Token token=\"$authn_azure_access_token\"" \
  https://"$CONJUR_SERVER_DNS":8443/secrets/cucumber/variable/secrets/test-variable)

echo "Retrieved secret ${secret} from Conjur!!!"
```
[Code reference](https://github.com/cyberark/conjur/blob/application_identity_validation/ci/authn-azure/run-authn-azure.sh#L16)

The requested situation is to support both response of current **Flattened** JWS JSON Serialization with content type `application/json` and encoded JWS **Compact** Serialization with content type `text\plain`. 

## Issue description
1. HTTP server best practices: Using `application\json` as the content type to encoded output (not a JSON) is bad practice.
    So encoded output shouldn't use `application/json` unless it will be in JSON format. 
    The `../autheticate` response is in JSON format so it makes sense to use `application\json` as the response content type.
        As well as `../logic`,  which returns the api key as `text\plain`.

> ## Solution - Accept HTTP Header

Reviewer: Comment from meeting:
If user supply accept: 'xml' he will get json
even if user supply accept 'json' he will get 'json' but it's not because he specify json it is because of default behavior, so we must specify that the only excepted header is "text\plain" in it's full qualifier name ("txt" will not be parsed as "text\plain" thus return json)

Author: Added

> ## Solution - Accept HTTP Header

Reviewer: Please add solutions that didn't passed: adding optional param etc, so it will be documented

Author: Added

Otherwise, returns json access token as `application/json`.

```ruby
def authenticate
    # get the authn access token
    content_type = (request.headers["Accept"] == "text\plain") ? :text : :json
    # encode Base64 if needed
    render content_type => auth_token
```
### Backwards compatibility
Default behaviour: Only if the `Accept` is set to `text\plain` the response will be encoded.

## Solution - URL Patten
According to [Rails Routing Guides](https://guides.rubyonrails.org/routing.html#route-globbing-and-wildcard-segments)  we can use wildcards to suggests the requested `Content_type`.

By requesting `../authenticate.encoded` the output will be encoded access token.    

### Design
When `../authenticate` request the response will be encoded **only** if the url suffix (wildcard) is `.encoded`.
Otherwise, returns json access token as `application/json`

```ruby
def authenticate
    # get the authn access token
    # extract the url suffix
    content_type = ( url_suffix == ".encoded") ? :text : :json
    # encode Base64 if needed
    render content_type => auth_token
```
### Backwards compatibility

> Default behaviour: If the request doesn't include the `encoded` wildcard explicitly the response will be decoded.

Reviewer: Current behavior is that `authenticate.*` returns 404

Author: Added

## Preferred Solution
Even though, rails routing guidelines suggests that it supports response's content type based on wildcard suffix,
I prefer the `Accept` HTTP header solution. This solution is is more accurate from content negotiation point of view.
The suffix based solution is less complicated but it makes the request seems like it ask for a file (which ends with a 
certain suffix) rather than a dynamic content type handling. The `Accept` method is more like a contract between the client 
and server - in our case the client accepts `text\plain` response causes the server to send an encoded token. 

## Security

> TBD

Reviewer: Opening open in Rails to suffix - how it affect us

Author: Added

## Test Plan

| **Title** | **Given** | **When** | **Then** | **Comment** |
|-----------|-----------|----------|----------|-------------|
| Successful encoded token response           | Credentials          | I send an authenticate request          |  I get an encoded valid token        | Authenticate using the chosen method            |
| Failed authenticate request works          | Wrong credentials          | I send an authenticate request         | I get the same response as decode flow         |             |

## Documentation
- [ ] Add documentation about new API ability

> ## Open questions

Reviewer: Performance degrigation, need to be verified

Author: Added

  * [Security](#security)
  * [Test Plan](#test-plan)
  * [Documentation](#documentation)
  * [Open questions](#open-questions)

## Glossary

| **Term** | **Description**  |
|----------|------------------|
| encoded  |base64 encoded    |
| JWS      |json web signature|
| JWT      |json web token    |

## Useful links
1. [HTTP Content Negotiation Best Practices](https://restfulapi.net/content-negotiation/)
2. MDN web docs:
    1. [HTTP Accept Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept)
    2. [HTTP Accept-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding)
    3. [HTTP Content-Type Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)
    4. [HTTP Content-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding)
3. [Rails Routing Guides](https://guides.rubyonrails.org/routing.html#route-globbing-and-wildcard-segments)
4. [Correct Usage of JSON](https://stackoverflow.com/questions/477816/what-is-the-correct-json-content-type)

## Background

As a Conjur User or Host, I want to send authentication request to `authn` authenticator (for example).

> So that I receive a JWT access token.

Reviewer 2: i would emphasize that you want a "ready-to-use" base64 encoded Conjur access token instead of needing to encode it yourself

Reviewer 2: and i think that the "As a" language is abit hard to understand here. I would state the current behaviour (currently, when a customer sends an authn request...), it's problem and the desired behaviour

Reviewer 2: i now see that you already wrote it later. in that case i think the whole sentence is redundant

Author: Changed

![alt text](encoded_token_chart.jpeg "Login process")

Current authentication response is a Flattened JWS JSON Serialization, e.g:
```json
{
    "protected": "eyJh...",
    "payload": "eyJz...",
    "signature": "CZgB..."
}
``` 

In order to use this token for restricted endpoint it **must** be encode and added to `Authorization: Token token=` header:
```sh
echo "Get Conjur access token using an Azure access token"
# Get an authn-azure Conjur access token for host azure-apps/test-app
authn_azure_response=$(curl -k -X POST \
  -H "Content-Type: application/x-www-form-urlencoded" \
  --data "jwt=$azure_access_token" \
  https://"$CONJUR_SERVER_DNS":8443/authn-azure/test/cucumber/host%2Fazure-apps%2Ftest-app/authenticate)

> authn_azure_access_token=$(echo -n "$authn_azure_response" | base64 | tr -d '\r\n')

Reviewer 2: i would add a blank line before this and emphasize that this is the bad ux

echo "Retrieve a secret using the Conjur access token"
# Retrieve a Conjur secret using the authn-azure Conjur access token
secret=$(curl -k -H "Authorization: Token token=\"$authn_azure_access_token\"" \
  https://"$CONJUR_SERVER_DNS":8443/secrets/cucumber/variable/secrets/test-variable)

echo "Retrieved secret ${secret} from Conjur!!!"
```
[Code reference](https://github.com/cyberark/conjur/blob/application_identity_validation/ci/authn-azure/run-authn-azure.sh#L16)

The requested situation is to support both response of current **Flattened** JWS JSON Serialization with content type `application/json` and encoded JWS **Compact** Serialization with content type `text\plain`. 

## Issue description

> 1. HTTP server best practices: Using `application\json` as the content type to encoded output (not a JSON) is bad practice.

Reviewer 3: you got here reversed slashes 'application/json` vs 'application\json`

    So encoded output shouldn't use `application/json` unless it will be in JSON format. 
    The `../autheticate` response is in JSON format so it makes sense to use `application/json` as the response content type.

>         As well as `../logic`,  which returns the api key as `text/plain`.

Reviewer 2: `../login`?

## Solution - Accept HTTP Headers
According to [MDN web docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept):
 > The `Accept` request HTTP header advertises which content types the client is able to understand. Using content negotiation, the server then selects one of the proposals, uses it and informs the client of its choice with the `Content-Type` response header.

To conclude, we can choose the response's `Content-Type` by the request's `Accept` header.

We can add the usage of `Accept-Encoding` HTTP header, which advertises which content type encoding the client is able to undetstand. 

### Design

> When `../authenticate` request the response will be encoded **only** if the `Accept` header is `text/plain` and `Accept-Encoding` is `base64`.

Reviewer 3: @liavyona why do we need both? is the Accept header you suggest here is not the "default" one?
can we make do with only "Accept-Encoding"?

Author: a. Yes we can (B. Obama).
b. The two headers combined together make are more certain to confirm the user actually asked for encoded response

Author: I can add this to the open question and we discuss it on the presentation

Reviewer 2: i agree with @InbalZilberman . if the user can add only one header to get the desired solution then it's better.

Reviewer 2: i would go with `-H "Accept-Encoding: base64"`. it makes sense and is clear.

## Omitted Solutions
* **Optional Parameters** - By passing an optional parameter (`../authenticate?encoded=true` e.g.) which requires 
an encoded response. This solution is not popular in Conjur because the server doesn't use url optional parameters.  

* **New Endpoint** - By requesting to a new endpoint (`../authenticate/encoded` e.g.) the response will be encoded. This 
solution is a redundant overhead which can (and should) be avoided. 

## Preferred Solution
Even though, rails routing guidelines suggests that it supports response's content type based on wildcard suffix,
I prefer the `Accept` HTTP header solution. This solution is is more accurate from content negotiation point of view.
The suffix based solution is less complicated but it makes the request seems like it ask for a file (which ends with a 
certain suffix) rather than a dynamic content type handling. The `Accept` method is more like a contract between the client 
and server - in our case the client accepts `text/plain` response causes the server to send an encoded token. 

## Security
TBD

## Test Plan

> 1. Tests should be on DAP and Conjur (HTTPS vs HTTP).

Reviewer 2: should we test this for each authenticator or is it enough to test it with `authn`?

cc: @eladkug

  * [Open questions](#open-questions)

## Glossary

| **Term** | **Description**  |
|----------|------------------|
| encoded  |base64 encoded    |
| JWS      |json web signature|
| JWT      |json web token    |

## Useful links
1. [HTTP Content Negotiation Best Practices](https://restfulapi.net/content-negotiation/)
2. MDN web docs:
    1. [HTTP Accept Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept)
    2. [HTTP Accept-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Encoding)
    3. [HTTP Content-Type Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type)
    4. [HTTP Content-Encoding Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Encoding)
3. [Rails Routing Guides](https://guides.rubyonrails.org/routing.html#route-globbing-and-wildcard-segments)
4. [Correct Usage of JSON](https://stackoverflow.com/questions/477816/what-is-the-correct-json-content-type)

## Background

When a Conjur identity sends authentication request to `authn` authenticator (for example), it receives a JWT access token.
An example for optional HTTP session flow: 

> ![alt text](encoded_token_chart.jpeg "Login process")

Reviewer 4: image isn't rendering

Author: https://github.com/cyberark/conjur/blob/19d7615a8088dd905222737ffc88f13523a0a180/design/proposals/encoded_token_response.md
